GraphBase/src/graph/UnidirectedGraph.java

```diff
@@ -0,0 +1,74 @@
+package graph;
+import java.lang.Cloneable;
+import java.util.Collection;
+import java.util.concurrent.Callable;
+
+public class UnidirectedGraph extends Graph implements Cloneable{
+
+  public Vertex deleteVertex(int vertexIndex) {
+
+    if (!vertices.containsKey(vertexIndex)) {
+      return null;
+    }
+    for (Integer key : vertices.get(vertexIndex).getAdjEdges().keySet()) {
+      this.deleteEdge(vertices.get(vertexIndex).getAdjEdges().get(key));
+    }
+    return vertices.remove(vertexIndex);
+  }
+
+  public Vertex deleteVertex(Vertex vertex) {
+    return deleteVertex(vertex.getId());
+  }
+
+  public boolean addEdge(int fromId, int toId) {
+    Edge newEdge = new Edge(fromId, toId);
+    return addEdge(newEdge);
+  }
+
+  public boolean addEdge(Vertex fromVertex, Vertex toVertex) {
+    return addEdge(fromVertex.getId(), toVertex.getId());
+  }
+
+  public boolean addEdge(Edge edge) {
+    addVertex(edge.getFromId());
+    addVertex(edge.getToId());
+    if (vertices.get(edge.getFromId()).isAdjacent(edge.getToId())) {
+      return false;//represents the edge has already existed.
+    }
+    vertices.get(edge.getFromId()).addFromAdjEdge(edge);
+    vertices.get(edge.getToId()).addToAdjEdge(edge);
+    return true;
+  }
+
+  public boolean addEdge(Collection<Edge> edgeCollection) {
+    for (Edge edge : edgeCollection) {
+      if (!addEdge(edge)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  public Edge deleteEdge(int fromId, int toId) {
+    vertices.get(fromId).deleteAdjEdge(toId);
+    return vertices.get(toId).deleteAdjEdge(fromId);
+  }
+
+  public Edge deleteEdge(Vertex fromVertex, Vertex toVertex) {
+    return deleteEdge(fromVertex.getId(), toVertex.getId());
+  }
+
+  public Edge deleteEdge(Edge edge) {
+    return deleteEdge(edge.getFromId(), edge.getToId());
+  }
+
+  @Override
+  public UnidirectedGraph clone() {
+    Collection<Vertex> vertexCollection = verticesCollection();
+    UnidirectedGraph newGraph = new UnidirectedGraph();
+    for (Vertex vertex : vertexCollection) {
+      newGraph.addVertex(vertex.clone());
```

Comment:
We have a problem here right? Suppose there is an edge (1, 2), in your current implementation, you must have the same edge object in 1 and 2's adjacency list. However, if you simply use this cloning, you will have different edge objects in the cloned graph.